File: rxbinding-design/src/main/java/com/jakewharton/rxbinding2/support/design/widget/TabLayoutSelectionEventObservable.java
package com.jakewharton.rxbinding2.support.design.widget;

import android.support.design.widget.TabLayout;
import android.support.design.widget.TabLayout.Tab;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread;

final class TabLayoutSelectionEventObservable extends Observable<TabLayoutSelectionEvent> {
  private final TabLayout view;

  TabLayoutSelectionEventObservable(TabLayout view) {
    this.view = view;
  }

  @Override protected void subscribeActual(Observer<? super TabLayoutSelectionEvent> observer) {
    if (!checkMainThread(observer)) {
      return;
    }
    Listener listener = new Listener(view, observer);
    observer.onSubscribe(listener);
    view.addOnTabSelectedListener(listener);

    int index = view.getSelectedTabPosition();
    if (index != -1) {
      observer.onNext(TabLayoutSelectionSelectedEvent.create(view, view.getTabAt(index)));
    }
  }

  final class Listener extends MainThreadDisposable implements TabLayout.OnTabSelectedListener {
    private final TabLayout tabLayout;
    private final Observer<? super TabLayoutSelectionEvent> observer;

    Listener(TabLayout tabLayout, Observer<? super TabLayoutSelectionEvent> observer) {
      this.tabLayout = tabLayout;
      this.observer = observer;
    }

    @Override public void onTabSelected(Tab tab) {
      if (!isDisposed()) {
        observer.onNext(TabLayoutSelectionSelectedEvent.create(view, tab));
      }
    }

    @Override public void onTabUnselected(Tab tab) {
      if (!isDisposed()) {
        observer.onNext(TabLayoutSelectionUnselectedEvent.create(view, tab));
      }
    }

    @Override public void onTabReselected(Tab tab) {
      if (!isDisposed()) {
        observer.onNext(TabLayoutSelectionReselectedEvent.create(view, tab));
      }
    }

    @Override protected void onDispose() {
      tabLayout.removeOnTabSelectedListener(this);
    }
  }
}

File: rxbinding-material/src/main/java/com/jakewharton/rxbinding3/material/TabLayoutSelectionEventObservable.kt
@file:JvmName("RxTabLayout")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.material

import androidx.annotation.CheckResult
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayout.Tab
import com.jakewharton.rxbinding3.internal.checkMainThread
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

/**
 * Create an observable which emits selection, reselection, and unselection events for the tabs
 * in `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 *
 * *Note:* If a tab is already selected, an event will be emitted immediately on subscribe.
 */
@CheckResult
fun TabLayout.selectionEvents(): Observable<TabLayoutSelectionEvent> {
  return TabLayoutSelectionEventObservable(this)
}

private class TabLayoutSelectionEventObservable(
  val view: TabLayout
) : Observable<TabLayoutSelectionEvent>() {

  override fun subscribeActual(observer: Observer<in TabLayoutSelectionEvent>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    observer.onSubscribe(listener)
    view.addOnTabSelectedListener(listener)

    val index = view.selectedTabPosition
    if (index != -1) {
      observer.onNext(TabLayoutSelectionSelectedEvent(view, view.getTabAt(index)!!))
    }
  }

  private class Listener(
    private val view: TabLayout,
    private val observer: Observer<in TabLayoutSelectionEvent>
  ) : MainThreadDisposable(), TabLayout.OnTabSelectedListener {

    override fun onTabSelected(tab: Tab) {
      if (!isDisposed) {
        observer.onNext(TabLayoutSelectionSelectedEvent(view, tab))
      }
    }

    override fun onTabUnselected(tab: Tab) {
      if (!isDisposed) {
        observer.onNext(TabLayoutSelectionUnselectedEvent(view, tab))
      }
    }

    override fun onTabReselected(tab: Tab) {
      if (!isDisposed) {
        observer.onNext(TabLayoutSelectionReselectedEvent(view, tab))
      }
    }

    override fun onDispose() {
      view.removeOnTabSelectedListener(this)
    }
  }
}
